doc/mkdocs/docs/api/basic_json/value.md

```diff
@@ -2,8 +2,8 @@
 
 ```cpp
 // (1)
-template<class ValueType>
-ValueType value(const typename object_t::key_type& key,
+template<class KeyType, class ValueType>
+ValueType value(const KeyType& key,
```

Comment:
Wouldn't you want this to be KeyType&& ?

Reply:
If I change the signature to

```cpp
ValueType value(KeyType&& key, const ValueType& default_value) const
```

and use

```cpp
const auto it = find(std::forward<KeyType>(key));
```

then I get the following error:

```
In file included from ../test/src/unit-element_access2.cpp:32:
../include/nlohmann/json.hpp:4046:16: error: no matching member function for call to 'value'
        return value(key, string_t(default_value));
               ^~~~~
../test/src/unit-element_access2.cpp:276:29: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::value<std::__1::basic_string_view<char, std::__1::char_traits<char> >, 0>' requested here
                    CHECK(j.value(std::string_view("string"), "bar") == "hello world");
                            ^
../include/nlohmann/json.hpp:4044:14: note: candidate function template not viable: no known conversion from 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::string_t' (aka 'std::__1::basic_string<char>') to 'const char *' for 2nd argument
    string_t value(const KeyType& key, const char* default_value) const
             ^
../include/nlohmann/json.hpp:4094:15: note: candidate function template not viable: no known conversion from 'const std::__1::basic_string_view<char, std::__1::char_traits<char> >' to 'const nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::json_pointer' (aka 'const json_pointer<nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >') for 1st argument
    ValueType value(const json_pointer& ptr, const ValueType& default_value) const
              ^
../include/nlohmann/json.hpp:4118:14: note: candidate function not viable: no known conversion from 'const std::__1::basic_string_view<char, std::__1::char_traits<char> >' to 'const nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::json_pointer' (aka 'const json_pointer<nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >') for 1st argument
    string_t value(const json_pointer& ptr, const char* default_value) const
             ^
../include/nlohmann/json.hpp:4017:15: note: candidate template ignored: requirement 'detail::is_getable<nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::value && !std::is_same<nlohmann::detail::value_t, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::value && (std::is_same<const std::__1::basic_string_view<char, std::__1::char_traits<char> > &, std::__1::basic_string_view<char, std::__1::char_traits<char> > >::value || std::is_convertible<const std::__1::basic_string_view<char, std::__1::char_traits<char> > &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::value)' was not satisfied [with KeyType = const std::__1::basic_string_view<char, std::__1::char_traits<char> > &, ValueType = std::__1::basic_string<char>]
    ValueType value(KeyType&& key, const ValueType& default_value) const
              ^
1 error generated.
```

Reply:
Well, you'd probably want the second parameter to be && as well, but that's not really relevant to your issue.

Maybe try adding an std::enable_if for if the KeyType converts to json_pointer, to avoid the ambiguity?

Reply:
There's no ambiguity. The compiler does not find a matching function.